### PR TITLE
feat: support VictoryChart

### PR DESCRIPTION
**Pull request by situchan**
**ExpensiMark:**
- Added **VictoryChart passthrough block** — single regex covers all 9 tags (`VictoryChart|VictoryBar|VictoryLine|VictoryPie|VictoryAxis|VictoryLabel|VictoryLegend|VictoryTooltip|VictoryGroup`) via alternation, case-insensitive (handles both `<VictoryChart>` from the server and lowercase `<victorychart>` from react-native-render-html).
- `extractPassthroughTags` / `restorePassthroughTags` helpers swap chart blocks for `EXMK_PT_<i>` placeholders before processing, restore after — same pattern as the` <video>` preservation, just generalized to a block rather than a single line.
- Wired into both `replace` (before escape) and `htmlToMarkdown` (before `replaceBlockElementWithNewLine`'s `stripHTML`).
Added `victoryChart` rule to `htmlToTextRules` emitting `[chart]`.

**Tests:**
- 11 round-trip tests (`htmlToMarkdown` ↔ `replace`), plus edge cases (lowercase, self-closing, all 9 inner tags, surrounding markdown rules, security invariant, stray placeholder text).

### Fixed Issues
$ https://github.com/Expensify/App/issues/90547

# Tests
1. What unit/integration tests cover your change? What autoQA tests cover your change?
1. What tests did you perform that validates your changed worked?

# QA
1. What does QA need to do to validate your changes?
1. What areas to they need to test for regressions?

### CODE REVIEW

**Comment by situchan** [comment]
@codex review

**Comment by luacmartins** [comment]
@situchan what's the ETA to get this in review?

**Comment by situchan** [comment]
Ready for review

**Comment by luacmartins** [comment]
This is part of the embedded charts project, so I'll review

File: lib/ExpensiMark.ts [hunk 1]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+// Server-rendered VirtualCFO charts are embedded in `reportAction.message.html` as a single
+// <VictoryChart> block. We preserve that whole block verbatim through `htmlToMarkdown` (and the
+// trusted-input path of `replace`) so the chart round-trips without its inner markup being
+// escaped or stripped. Matching only the outer <VictoryChart> container is deliberate:
+//  - every other Victory tag (VictoryBar/Line/Pie/Axis/Label/Legend/Tooltip/Group) is always
+//    nested inside it, so it is preserved as part of the captured block - no separate rule needed;
+//  - it avoids the unbalanced-nesting failure a lazy `[\s\S]*?` would hit if a self-nesting tag
+//    such as <VictoryGroup> were matched on its own.
+// Case-insensitive matching covers both the PascalCase the server emits (<VictoryChart>) and the
+// lowercase form react-native-render-html uses internally (<victorychart>).
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
// Preserve VirtualCFO chart blocks by matching the outer <VictoryChart> container.
// This captures all nested Victory components and prevents markup escaping during conversion.
```

File: lib/ExpensiMark.ts [hunk 2]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+const PASSTHROUGH_TAGS_REGEX = /<VictoryChart\b[^>]*\/>|<VictoryChart\b[^>]*>[\s\S]*?<\/VictoryChart>/gi;
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
const VICTORY_CHART_REGEX = /<VictoryChart\b[^>]*\/>|<VictoryChart\b[^>]*>[\s\S]*?<\/VictoryChart>/gi;
```

File: lib/ExpensiMark.ts [hunk 3]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+// Placeholders are wrapped in a NUL sentinel. NUL never occurs in real HTML/markdown text, so a
+// chart placeholder can never collide with user-typed content, and the numeric core carries no
+// markdown-significant characters that other rules could mangle. Built via String.fromCharCode
+// so no control character is ever stored literally in this source file.
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion

// Chart placeholders use NUL characters as delimiters to prevent conflicts with user content.
```

File: lib/ExpensiMark.ts [hunk 4]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+const PASSTHROUGH_SENTINEL = String.fromCharCode(0);
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
const VICTORY_CHART_PLACEHOLDER_DELIMITER = String.fromCharCode(0);
```

File: lib/ExpensiMark.ts [hunk 5]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+const PASSTHROUGH_PLACEHOLDER_REGEX = new RegExp(`${PASSTHROUGH_SENTINEL}(\\d+)${PASSTHROUGH_SENTINEL}`, 'g');
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
const VICTORY_CHART_PLACEHOLDER_PATTERN = new RegExp(`${VICTORY_CHART_PLACEHOLDER_DELIMITER}(\\d+)${VICTORY_CHART_PLACEHOLDER_DELIMITER}`, 'g');
```

File: lib/ExpensiMark.ts [hunk 6]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+const buildPassthroughPlaceholder = (index: number): string => `${PASSTHROUGH_SENTINEL}${index}${PASSTHROUGH_SENTINEL}`;
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
const createVictoryChartPlaceholder = (index: number): string => `${VICTORY_CHART_PLACEHOLDER_DELIMITER}${index}${VICTORY_CHART_PLACEHOLDER_DELIMITER}`;
```

File: lib/ExpensiMark.ts [hunk 7]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+
 export default class ExpensiMark {
+    extractPassthroughTags = (text: string): {text: string; tags: string[]} => {
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
    extractVictoryChartTags = (text: string): {text: string; tags: string[]} => {
```

File: lib/ExpensiMark.ts [hunk 8]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+        const tags: string[] = [];
+        const out = text.replace(PASSTHROUGH_TAGS_REGEX, (match) => {
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
        const out = text.replace(VICTORY_CHART_REGEX, (match) => {
```

File: lib/ExpensiMark.ts (hunk 7 above)
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
    victoryChartTags = (text: string): {text: string; tags: string[]} => {
```

File: lib/ExpensiMark.ts [hunk 9]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+            const placeholder = buildPassthroughPlaceholder(tags.length);
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
            const placeholder = createVictoryChartPlaceholder(tags.length);
```

File: lib/ExpensiMark.ts [hunk 10]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+            tags.push(match);
+            return placeholder;
+        });
+        return {text: out, tags};
+    };
+
+    restorePassthroughTags = (text: string, tags: string[]): string => {
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
    restoreVictoryChartTags = (text: string, tags: string[]): string => {
```

File: lib/ExpensiMark.ts [hunk 11]
```diff
@@ -83,7 +83,43 @@ const MARKDOWN_VIDEO_REGEX = new RegExp(
...
+        if (tags.length === 0) {
+            return text;
+        }
+        return text.replace(PASSTHROUGH_PLACEHOLDER_REGEX, (match, idx) => tags[Number(idx)] ?? match);
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
        return text.replace(VICTORY_CHART_PLACEHOLDER_PATTERN, (match, idx) => tags[Number(idx)] ?? match);
```

File: lib/ExpensiMark.ts [hunk 12]
```diff
@@ -968,8 +1009,18 @@ export default class ExpensiMark {
...
         }
 
+        // Lift out <VictoryChart> blocks before processing so neither escaping nor markdown rules
+        // disturb their inner markup (JSON in attributes, special chars, etc.). Restored at the end.
+        //
+        // This is gated on `!shouldEscapeText`: passthrough is only safe for trusted input (server
+        // HTML re-parsed with shouldEscapeText === false). When shouldEscapeText is true the input
+        // is user-typed markdown, and a chart tag must be escaped like any other HTML - otherwise a
+        // user could smuggle arbitrary markup (e.g. <VictoryChart><img onerror=...></VictoryChart>)
+        // straight past escaping.
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
        // Extract VictoryChart blocks to preserve their markup during processing.
        // Only safe for trusted server input - user input must be escaped to prevent XSS.
```

File: lib/ExpensiMark.ts [hunk 13]
```diff
@@ -1003,7 +1054,7 @@ export default class ExpensiMark {
             return shouldEscapeText ? Utils.escapeText(text) : text;
         }
 
-        return replacedText;
+        return this.restorePassthroughTags(replacedText, passthroughTags);
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
        return this.restoreVictoryChartTags(replacedText, passthroughTags);
```

File: lib/ExpensiMark.ts [hunk 14]
```diff
@@ -968,8 +1009,18 @@ export default class ExpensiMark {
...
 
+        // Lift out <VictoryChart> blocks before processing so neither escaping nor markdown rules
+        // disturb their inner markup (JSON in attributes, special chars, etc.). Restored at the end.
+        //
+        // This is gated on `!shouldEscapeText`: passthrough is only safe for trusted input (server
+        // HTML re-parsed with shouldEscapeText === false). When shouldEscapeText is true the input
+        // is user-typed markdown, and a chart tag must be escaped like any other HTML - otherwise a
+        // user could smuggle arbitrary markup (e.g. <VictoryChart><img onerror=...></VictoryChart>)
+        // straight past escaping.
+        const {text: textWithPlaceholders, tags: passthroughTags} = shouldEscapeText ? {text, tags: [] as string[]} : this.extractPassthroughTags(text);
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
        const {text: textWithPlaceholders, tags: victoryChartTags} = shouldEscapeText ? {text, tags: [] as string[]} : this.extractVictoryChartTags(text);
```

File: lib/ExpensiMark.ts [hunk 15]
```diff
@@ -1278,7 +1334,8 @@ export default class ExpensiMark {
         };
 
         this.htmlToMarkdownRules.forEach(processRule);
-        return Str.htmlDecode(this.replaceBlockElementWithNewLine(generatedMarkdown));
+        const decoded = Str.htmlDecode(this.replaceBlockElementWithNewLine(generatedMarkdown));
+        return this.restorePassthroughTags(decoded, passthroughTags);
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
        return this.restoreVictoryChartTags(decoded, victoryChartTags);
```

File: lib/ExpensiMark.ts [hunk 16]
```diff
@@ -1268,6 +1319,11 @@ export default class ExpensiMark {
         }
         generatedMarkdown = this.unpackNestedQuotes(generatedMarkdown);
 
+        // Lift out <VictoryChart> blocks before replaceBlockElementWithNewLine's stripHTML would
+        // erase them. Restored after rule processing so the chart fragment survives the round-trip.
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
        // Extract VictoryChart blocks before HTML stripping, then restore them.
```

File: lib/ExpensiMark.ts [hunk 17]
```diff
@@ -1268,6 +1319,11 @@ export default class ExpensiMark {
...
+        const {text: textWithPlaceholders, tags: passthroughTags} = this.extractPassthroughTags(generatedMarkdown);
```
Comment (luacmartins) [inline lib/ExpensiMark.ts]:
```suggestion
        const {text: textWithPlaceholders, tags: passthroughTags} = this.extractVictoryChartTags(generatedMarkdown);
```

File: __tests__/ExpensiMark-HTML-test.js [hunk 18]
```diff
@@ -2503,3 +2503,111 @@ describe('room mentions', () => {
         expect(parser.replace(testString)).toBe(resultString);
     });
 });
+
+// VirtualCFO charts are embedded directly into reportAction.message.html as <VictoryChart> blocks.
+// The chart fragment (and its inner Victory* tags) must survive htmlToMarkdown and the trusted-input
+// path of the markdown-to-html replace step so round-tripping doesn't strip or escape the markup.
```
Comment (luacmartins) [inline __tests__/ExpensiMark-HTML-test.js]:
```suggestion
// VirtualCFO charts must preserve their VictoryChart markup during HTML/markdown conversion.
```

**Review by luacmartins — CHANGES_REQUESTED** [review]

**Review by luacmartins — APPROVED** [review]
